### ISSUE TITLE
Setting checkbox value in template

### ISSUE THREAD

**Issue by ozilion** [PHPOffice/PHPWord#1187]
I think still this is one of the most wanted feature of Templates. Can we add this too?

**Comment by FBnil** [comment]
@ozilion Can be done already, but how do you imagine this function? Do you free-form search for some text, and find the first checkbox to the left of it? Or do you enumerate the checkboxes and search for the n-th one? The unique id of the w:sdt tag?

```
<w:sdt>
<w14:checkbox>
    <w14:checked w14:val="1"/>
:
         <w:t>☒</w:t>
      </w:r>
    </w:sdtContent>
</w:sdt>
```

**Comment by ozilion** [comment]
I tried PHPDocx, they use bookmark in checkbox. Defining bookmark variable for each checkbox should work or may be better solution? I have a forms contains many checkboxes in but depending some other criteria i just have to check 1 to 3 of them.

**Comment by ozilion** [comment]
By the way changing

<w14:checked w14:val="1"/>
to 
<w14:checked w14:val="0"/>

will work

**Comment by FBnil** [comment]
I thought, or at least [docx4j](https://github.com/plutext/docx4j/commit/e98ce9ab61f90b72cc616b77a03c72b54c8ca0bc) does, set both the ☒ AND the val="1" (or vice versa, text.setValue("☐"); and val="0"). 

Bookmarks, huh. Nice unobtrusive way to add variables. I'll see what I can do, give me a few days.

**Comment by FBnil** [comment]
Minimal working example. Requires [my fork of TemplateProcessor.php](https://github.com/FBnil/PHPWord/blob/FBnil-patch-blocks/src/PhpWord/TemplateProcessor.php) (drop-in replacement for 0.13.0/0.14.0)
@ozilion So yes, totally doable, but lacking in functionality. Do you have more suggestions to build up a forms library?
```
class CheckMarkTemplateProcessor extends \PhpOffice\PhpWord\TemplateProcessor {
	public function setCheckOn($macro){
		$xmldata = $this->processSegment(static::ensureMacroCompleted($macro), 'w:sdt', 1, 'MainPart', function(&$xmlSegment, &$segmentStart, &$segmentEnd, &$part){
			$closetag = '</w:sdt>';
			$segmentEnd = strpos($part, $closetag, $segmentStart + 1);
			if (!$segmentEnd) {
				$xmlSegment = null; # flag the error
				return false; # do not update the xml, return the $xmlSegment (null) 
			}
			$segmentEnd += strlen($closetag); # also include the closing tag in our segment
			$xmlSegment = substr($part, $segmentStart, ($segmentEnd - $segmentStart));
			$xmlSegment = str_replace(
				// MSWord uses spaces for closed tags, LibreOffice does not.
				[ '<w14:checked w:val="0"/>', '<w14:checked w:val="0" />', '<w:t>☐</w:t>' ],
				[ '<w14:checked w:val="1"/>', '<w14:checked w:val="1" />', '<w:t>☒</w:t>' ], 
				$xmlSegment
			);
			return $xmlSegment; # replace segment and return the $xmlSegment
		});

		if($xmldata === false){
			dd("FATAL: variable 'bookmark_1' not found");
		}elseif($xmldata === null){
			dd("FATAL: variable 'bookmark_1' found, but checkbox not found");
		}
		return $xmldata; # success
	}
}

$file_dir = storage_path('checkbox_withbookmark.docx');
$template = new CheckMarkTemplateProcessor($file_dir);
$template->setCheckOn('bookmark_1'); # searches for ${bookmark_1} and the checkbox to the left
$template->saveAs(storage_path('app/public/x.docx'));
return response()->download(storage_path('app/public/x.docx'))->deleteFileAfterSend(true);
```

**Comment by ozilion** [comment]
I think this is the one of the most expected feature (at least for me) and personally i could use convertion to pdf but it is another matter. In our project we only need now is to change state of any checkbox in templates. Thanks for your efforts, you have enough time :)

**Comment by FBnil** [comment]
Conversion to pdf is a whole new bag of cats, as it goes word->html->pdf and html markup is lacking if you want more than just the basics. If you need a perfect word to pdf, either get a commercial version or install the commandline version of open/libreoffice, which just works but is very slow and does not scale (or look into UNO if you need speed; search for "unoconv")

**Comment by FBnil** [comment]
@ozilion Ok, time for you to be my first tester.
From [my branch](https://github.com/FBnil/PHPWord/tree/templates)

grab 2 files:
[TemplateProcessor.php](https://github.com/FBnil/PHPWord/blob/templates/src/PhpWord/TemplateProcessor.php)
[CheckboxTemplateProcessor.php](https://github.com/FBnil/PHPWord/blob/templates/src/PhpWord/CheckboxTemplateProcessor.php)
and put them in your ./vendor/phpoffice/PHPWord/src/PhpWord/  directory.

From there (you can read the testcases and source to see what it does)

```php
$file_dir = storage_path('checkbox_withbookmark.docx');
$template = new \PhpOffice\PhpWord\CheckboxTemplateProcessor($file_dir);
$template->setCheckOn('bookmark_1');
$template->setCheckOff('bookmark_2');
$template->saveAs(storage_path('app/public/tmp.docx'));
return response()->download(storage_path('app/public/tmp.docx'))->deleteFileAfterSend(true);
```
Because CheckboxTemplateProcessor inherits all from my rewrite of TemplateProcessor, you have also the goodies that TemplateProcessor brings (in fact, we use its segments to do the hard work)
You can read the text in [my pull request](https://github.com/PHPOffice/PHPWord/pull/1147) to see what it can do.

I really wish PHPOffice will drop support for PHP5.3.3 soon, because it is better to use traits instead of subclasses. (because you can mix-and-match with traits).

**Comment by ozilion** [comment]
@FBnil I still use v12. Will your files work with me?

**Comment by ozilion** [comment]
V13 not working with me, it genereates corrupted files from templates. in v12 i can use it without problem but your files didn't work for me.

**Comment by ozilion** [comment]
I made it work v13 but $template->setCheckboxOn('bookmark_1'); didn't work. Checkbox still unchecked. But can i see your test word file, may i made a mistake in my template defining bookmark?

**Comment by FBnil** [comment]
The version does not matter much as you are replacing the full TemplateProcessor with something else.
I made sure that corruption should not be a problem anymore (so 0.13.0 with this patch should work fine), are you saying that corruption still occurs with my version of TemplateProcessor? Do you have a docx you can share with the problem?

What did not work? Does `$result=$template->setCheckboxOn('bookmark_1');` return false? Did you add the bookmark in MSWord as `${bookmark_1}` ? You can also use macro names, just add a macro in the text right to the checkbox.

You should have something like:
```xml
<w:p>
<w:pPr><w:pStyle w:val="Normal"/><w:rPr></w:rPr></w:pPr>
<w:sdt>
	<w:sdtPr>
		<w14:checkbox>
			<w14:checked w:val="0"/>
			<w14:checkedState w:val="2612"/>
			<w14:uncheckedState w:val="2610"/>
		</w14:checkbox>
	</w:sdtPr>
	<w:sdtContent>
		<w:r>
			<w:rPr><w:rFonts w:eastAsia="MS Gothic" w:ascii="MS Gothic" w:hAnsi="MS Gothic"/><w:lang w:val="en-US"/></w:rPr>
			<w:t>☐</w:t>
		</w:r>
	</w:sdtContent>
</w:sdt>
<w:r><w:rPr><w:lang w:val="en-US"/></w:rPr><w:t xml:space="preserve"> </w:t></w:r>
<w:bookmarkStart w:id="0" w:name="${bookmark_1}"/><w:bookmarkEnd w:id="0"/>
<w:r><w:rPr><w:lang w:val="en-US"/></w:rPr><w:t xml:space="preserve">This is </w:t></w:r>
<w:r><w:rPr><w:lang w:val="en-US"/></w:rPr><w:t>a unchecked checkbox</w:t></w:r>
<w:r><w:rPr><w:lang w:val="en-US"/></w:rPr><w:t xml:space="preserve"> </w:t></w:r>
<w:r><w:rPr><w:lang w:val="en-US"/></w:rPr><w:t>${line1}</w:t></w:r>
</w:p>
```

**Comment by FBnil** [comment]
Here is a [document you requested](https://github.com/FBnil/PHPWord/blob/templates/samples/resources/Sample_38_CheckboxTemplate.docx)

**Comment by ozilion** [comment]
[2.AFR-02BasvuruGozdenGecirme-R5.docx](https://github.com/PHPOffice/PHPWord/files/1478235/2.AFR-02BasvuruGozdenGecirme-R5.docx)
this is one of my template files,  trying to test with check box near 'Kuruluşta tasarım ve geliştirme olmaması'

At first, the created files cannot be opened it was saying corrupted or broken file. Then i managed to create files succesfully but now only my first of about 25 templates replaced but others remains as original template, variables not converting. New TemplateProcessor not working in v12. Lastly I cannot add bookmark with special characters like ${ and }, using Office Pro 365 Plus

**Comment by FBnil** [comment]
Ok, there are several ways to tackle this and you need to choose the most logical method.

Your document contains:

```xml
<w:fldChar w:fldCharType="begin">
  <w:ffData>
    <w:name w:val="tasarim73"/><w:enabled/><w:calcOnExit w:val="0"/>
    <w:statusText w:type="text" w:val="@tasarim73@"/>
    <w:checkBox><w:sizeAuto/><w:default w:val="0"/></w:checkBox>
  </w:ffData>
</w:fldChar>
```

We also have your bookmark (which NEEDS to be unique), which is mixed up with another bookmark.
```xml
<w:bookmarkStart w:id="1" w:name="tasarim73"/>
	<w:bookmarkStart w:id="2" w:name="_GoBack"/>
<w:bookmarkEnd w:id="1"/>
	<w:bookmarkEnd w:id="2"/>
```

and we have several copy-pasted, same name:
```xml
<w:fldChar w:fldCharType="begin">
  <w:ffData>
    <w:name w:val="Onay1"/><w:enabled/><w:calcOnExit w:val="0"/>
    <w:checkBox><w:sizeAuto/><w:default w:val="0"/></w:checkBox>
  </w:ffData>
</w:fldChar>
```

Now the LibreOffice implementation is totally different from the MSWord implementation. In fact, I can not manipulate a MSWord checkbox in LibreOffice, and reading your reply, editing a LibreOffice checkbox is not possible in MSWord. Sigh.

Setting your first checkbox, using the w:ffData w:name is done like this:

```php

$file_dir = storage_path('2.AFR-02BasvuruGozdenGecirme-R5.docx');
$template = new \PhpOffice\PhpWord\CheckboxTemplateProcessor($file_dir);
\PhpOffice\PhpWord\CheckboxTemplateProcessor::$ensureMacroCompletion = false;

// quick function that only uses TemplateProcessor
function setCheckboxMS(&$template, $search, $newValue=true){
	$template->processSegment(
		$search,
		'w:checkBox',
		\PhpOffice\PhpWord\CheckboxTemplateProcessor::SEARCH_LEFT,
		0,
		'MainPart',
		function (&$xmlSegment, &$segmentStart, &$segmentEnd, &$part) use ($newValue) {
			$newValue = $newValue? 1 : 0;
			$count = 0;
			$xmlSegment = preg_replace(
				[ '~<w:default w:val="[^"]+"(\s?)/>~u' ],
				[ '<w:default w:val="'.$newValue.'"\1/>' ],
				$xmlSegment,
				1,
				$count
			);
			return $xmlSegment; // replace segment and return the $xmlSegment
		}
	);	
}

setCheckboxMS($template, 'w:name="@tasarim73@"', true); // search on w:statusText
setCheckboxMS($template, 'w:val="tasarim73"', true); // search on w:ffData name value
setCheckboxMS($template, 'w:name="tasarim73"', true); // search on bookmark name
setCheckboxMS($template, 'Çalışanların', true); // search on text (careful, NOT guaranteed to work)
setCheckboxMS($template, 'w:val="Onay1"', true);
```

Also note that there is no guarantee that a text is continuous, it could be:
`<w:t xml:space="preserve">Kuruluşta tasarım </w:t>` and the rest of the text is 100 characters to the right, with xml in between. using `${macroname}` actually searches and find these (and removes all xml between the macro, sometimes corrupting the document just by opening it with TemplateProcessor). 
However, searching on bookmarks or uniquely named ffData should always work, as that is kept together by xml.

**Comment by FBnil** [comment]
note: TemplateProcessor has been updated yet again (bug with search direction, which we use, update recommended).

**Comment by ozilion** [comment]
Thank you very much for quick responses.
I think using <w:name w:val="tasarim73"/><w:enabled/> is better because it's also create bookmark when naming.

**Comment by ozilion** [comment]
Ok. In Microsoft Office <w:name w:val="tasarim73"/>section is bookmark of checkbox element and <w:statusText w:type="text" w:val="@tasarim73@"/>is help text of checkbox. Also Microsoft 
Office does not allow any special charackters in bookmark names.

Whatever we choose it should be fastest and guaranted way for setting checkbox.

If this 'setCheckboxMS($template, 'w:val="Onay1"', true);' will work, Onay1 can be use as defined name and be unique as well.

**Comment by ozilion** [comment]
Final update;
Using
`setCheckboxMS($template, 'w:name="tasarim73"', true);`
Worked like i need but to do it I set search direction to 0(around as it says in function head text). Here it is 
`function setCheckboxMS(&$template, $search, $newValue=true){
        $template->processSegment(
            $search,
            'w:checkBox',
            0,--> direction
            0,
            'MainPart',`

This way checkboxes i want set correctly and i used PhpWprd v12 with minor hacks. Your code should work directly in v13 without any changes. But this changes, broke other replaces which is, especially all my search values in my templates after chekbox been checked not replacing still remains as template.

**Comment by FBnil** [comment]
@ozilion You mean that variables like `@firmaadi@` are not replaced? You need to disable the macro completion, or else it will search for `${@firmaadi@}`:
 
```php
\PhpOffice\PhpWord\TemplateProcessor::$ensureMacroCompletion = false;
```
(or `\PhpOffice\PhpWord\CheckboxTemplateProcessor` if you are still using that subclass)

**Comment by ozilion** [comment]
Any news about this to be added to new versions as default feature?

**Comment by Progi1984** [comment]
Fixed by #2509
